util/onion/src/onion_service.rs

```diff
@@ -0,0 +1,215 @@
+
+        let mut tor_controller = TorController::new(tor_controller, tor_password, None).await?;
+
+        tor_controller.wait_tor_server_bootstrap_done().await?;
+
+        let p2p_listener_addresses = [(
+            self.config.p2p_listen_address.port(),
+            self.config.p2p_listen_address,
+        )];
+        info!(
+            "Adding onion service v3: {}",
+            self.config.p2p_listen_address.to_string()
+        );
+        tor_controller
+            .add_onion_v3(self.key.clone(), &mut p2p_listener_addresses.iter())
+            .await
+            .map_err(|err| {
+                InternalErrorKind::Other.other(format!("Failed to add onion service: {:?}", err))
+            })?;
+        info!(
+            "Added onion service v3: {}",
+            self.config.p2p_listen_address.to_string()
+        );
+
+        self.handle.spawn(async move {
+            let mut ticker = tokio::time::interval(tokio::time::Duration::from_secs(3));
+            loop {
+                tokio::select! {
+                    _ = ticker.tick() => {
+                        let uptime = tor_controller.get_uptime().await;
+                        if let Err(err) = uptime {
+                            error!("Failed to get tor server uptime: {:?}", err);
+                            drop(tor_server_alive_tx);
+                            return;
+                        }
+                    }
+                    _ = stop_rx.cancelled() => {
+                        info!("OnionService received stop signal, exiting...");
+                        drop(tor_server_alive_tx);
+                        return;
+                    }
+                }
+            }
+        });
+        Ok(())
+    }
+}
+
+fn load_or_create_tor_secret_key(onion_private_key_path: String) -> Result<TorSecretKeyV3, Error> {
+    let is_onion_private_key_exists = Path::new(&onion_private_key_path).exists();
+    let key = match is_onion_private_key_exists {
+        true => load_tor_secret_key(onion_private_key_path)?,
+        false => create_tor_secret_key(onion_private_key_path)?,
+    };
+    Ok(key)
+}
+
+fn create_tor_secret_key(onion_private_key_path: String) -> Result<TorSecretKeyV3, Error> {
+    let key = torut::onion::TorSecretKeyV3::generate();
+    info!(
+        "Generated new onion service v3 key for address: {}",
+        key.public().get_onion_address()
+    );
+
+    let mut options = OpenOptions::new();
+    let mut options = options.create(true).truncate(true).write(true);
+
+    #[cfg(unix)]
+    {
+        use std::os::unix::fs::OpenOptionsExt;
+        options = options.mode(0o600);
+    }
+
+    let mut file = options.open(&onion_private_key_path).map_err(|err| {
+        InternalErrorKind::Other.other(format!(
+            "Failed to open onion private key for writing: {:?}",
+            err
+        ))
+    })?;
+    file.write_all(
+        &base64::engine::general_purpose::STANDARD
```

Comment:
can we write the key as binary directly to the file, why need base64 encoding/decoding here?